onnx/defs/reduction/defs.cc

```diff
@@ -156,37 +162,41 @@ The type of the output tensor is integer.)DOC";
         OpSchema::all_numeric_types(),
         "Constrain input and output types to all numeric tensors.");
     schema.TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+      // Type inference
       // set output element type to int64
       updateOutputElemType(ctx, 0, TensorProto_DataType_INT64);
 
+      // Shape inference
       if (!hasNInputShapes(ctx, 1)) {
         return;
       }
 
-      auto& input_shape = ctx.getInputType(0)->tensor_type().shape();
+      const auto& input_shape = ctx.getInputType(0)->tensor_type().shape();
       auto output_shape =
           ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape();
-      int64_t input_ndim = input_shape.dim_size();
+      const int64_t input_ndim = input_shape.dim_size();
       int64_t axis = 0; // default to 0
-      auto axis_proto = ctx.getAttribute("axis");
+      const auto* axis_proto = ctx.getAttribute("axis");
       if (axis_proto) {
         axis = axis_proto->i();
         if (axis < 0)
           axis += input_ndim;
+        if (axis < -input_ndim || axis >= input_ndim)
```

Comment:
```suggestion
        if (axis < 0 || axis >= input_ndim)
```

Reply:
I think that "axis" attribute is used so widely, I would recommend adding a utility function that adjusts negative axes, checks it has a valid value, throws an exception otherwise, and calling it. E.g., something like "getAxisAttribute(ctx, "axis", rank)".